backend.go

```diff
@@ -149,10 +151,57 @@ type backend struct {
 	checkOutLocks []*locksutil.LockEntry
 }
 
+// walkfunc type takes a storage path argument and returns true if a storage
+// entry exists, false otherwise
+type walkFunc func(string) (bool, error)
+
+// walkStoragePath performs a non-recursive breadth-first search of the given
+// storage path and applies the walkfunc to each storage entry
+func walkStoragePath(ctx context.Context, s logical.Storage, path string, walker walkFunc) error {
+	keys, err := s.List(ctx, path)
+	if err != nil {
+		return fmt.Errorf("unable to list keys: %w", err)
+	}
+
+	// Storage entries can be defined with hierarchical paths, e.g.
+	// "foo/bar/baz". But the storage.List() call to the top-level key will
+	// only return the top-level keys in the hierarchy. So we perform a
+	// non-recursive breadth-first search through all the keys returned from
+	// storage.List() and apply the walkfunc.
+	for i := 0; i < len(keys); i++ {
+		key := keys[i]
+
+		entryExists, err := walker(key)
+		if err != nil {
+			return fmt.Errorf("unable to read entry %q: %w", key, err)
+		}
+
+		if !entryExists && strings.HasSuffix(key, "/") {
+			// this is a directory
+			subKeys, err := s.List(ctx, path+key)
+			if err != nil {
+				return fmt.Errorf("unable to list keys: %w", err)
```

Comment:
Is there ever a case where "foo/bar" would fail, but the "foo/bar/baz" would succeed and both of those would be "correct"?

Reply:
An error returned from `storage.List()` means we could not access storage or we are in some bad state with the keys and we are dead in the water.

We have tests to cover the cases that I think you are describing:
https://github.com/hashicorp/vault-plugin-secrets-openldap/blob/156b280cd38c4d77e61d34839dad42396f9384f1/rotation_test.go#L48-L55